File: 30 Days of Code/C#/25 - Day 24 - More Linked Lists.cs
// ========================
//       Information
// ========================

// Direct Link: https://www.hackerrank.com/challenges/30-linked-list-deletion/problem
// Difficulty: Easy
// Max Score: 30
// Language: C#

// ========================
//         Solution
// ========================

using System;
using System.Collections.Generic;

class Node {
	public int data;
	public Node next;
	public Node(int d) {
		data = d;
		next = null;
	}
}

class Solution {
	public static Node removeDuplicates(Node head) {
		//Write your code here
		Node start = head;

		while (start.next != null) {
			if (start.data == start.next.data) {
				start.next = start.next.next;
			} else {
				start = start.next;
			}
		}
		return head;
	}

	public static Node insert(Node head, int data) {
		Node p = new Node(data);

		if (head == null) {
			head = p;
		}
		else if (head.next == null) {
			head.next = p;
		}
		else {
			Node start = head;
			while (start.next != null)
			start = start.next;
			start.next = p;

		}
		return head;
	}
	public static void display(Node head) {
		Node start = head;
		while (start != null) {
			Console.Write(start.data + " ");
			start = start.next;
		}
	}

	static void Main(String[] args) {
		Node head = null;

		int T = Int32.Parse(Console.ReadLine());
        
		while (T-->0) {
			int data = Int32.Parse(Console.ReadLine());
			head = insert(head, data);
		}
		head = removeDuplicates(head);
		display(head);
	}
}

File: 30 Days of Code/Python/25 - Day 24 - More Linked Lists.py
# ========================
#       Information
# ========================

# Direct Link: https://www.hackerrank.com/challenges/30-linked-list-deletion/problem
# Difficulty: Easy
# Max Score: 30
# Language: Python

# ========================
#         Solution
# ========================

class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
class Solution:
    def insert(self, head, data):
        p = Node(data)
        if head is None:
            head = p
        elif head.next is None:
            head.next = p
        else:
            start = head
            while start.next is not None:
                start = start.next
            start.next = p
        return head
    def display(self, head):
        current = head
        while current:
            print(current.data, end=' ')
            current = current.next

    def removeDuplicates(self, head):
        #Write your code here
        if head is None:
            return head
        values = [head.data]
        temp = head.next
        prev = head
        while temp:
            if temp.data in values:
                temp = temp.next
            else:
                values.append(temp.data)
                prev.next = temp
                prev = prev.next
                temp = temp.next
        prev.next = None
        return head

mylist= Solution()
T = int(input())
head = None
for i in range(T):
    data = int(input())
    head = mylist.insert(head, data)
head = mylist.removeDuplicates(head)
mylist.display(head)
